src/coreclr/nativeaot/System.Private.CoreLib/src/System/Runtime/InteropServices/TypeMapLazyDictionary.NativeAot.cs

```diff
@@ -0,0 +1,218 @@
+
+                ExternalReferencesTable externalReferences = default;
+                externalReferences.InitializeCommonFixupsTable(module);
+
+                var lookup = externalTypeMapTable.Lookup(typeMapGroupHandle.GetHashCode());
+                NativeParser entryParser;
+                while (!(entryParser = lookup.GetNext()).IsNull)
+                {
+                    RuntimeTypeHandle foundTypeMapGroup = externalReferences.GetRuntimeTypeHandleFromIndex(entryParser.GetUnsigned());
+                    if (!foundTypeMapGroup.Equals(typeMapGroupHandle))
+                    {
+                        continue;
+                    }
+                    bool isValid = entryParser.GetUnsigned() == 1;
+                    if (!isValid)
+                    {
+                        unsafe
+                        {
+                            delegate*<void> exceptionStub = (delegate*<void>)externalReferences.GetFunctionPointerFromIndex(entryParser.GetUnsigned());
+                            exceptionStub();
+                            Debug.Fail("Expected exception stub to throw an exception.");
+                            return null; // Should never reach here, as the exception stub should throw an exception.
+                        }
+                    }
+
+                    return new AssociatedTypeMapDictionary(new NativeHashtable(entryParser), externalReferences);
+                }
+            }
+
+            return new Dictionary<Type, Type>(); // Return an empty dictionary if no valid type map is found.
+        }
+
+        private static unsafe bool TryGetNativeReaderForBlob(TypeManagerHandle module, ReflectionMapBlob blob, out NativeReader reader)
+        {
+            byte* pBlob;
+            uint cbBlob;
+
+            if (RuntimeImports.RhFindBlob(module, (uint)blob, &pBlob, &cbBlob))
+            {
+                reader = new NativeReader(pBlob, cbBlob);
+                return true;
+            }
+
+            reader = default;
+            return false;
+        }
+
+        private class ExternalTypeMapDictionary(NativeHashtable table, ExternalReferencesTable externalReferences) : IReadOnlyDictionary<string, Type>
+        {
+            public Type this[string key]
+            {
+                get
+                {
+                    if (!TryGetValue(key, out Type? value))
+                    {
+                        ThrowHelper.ThrowKeyNotFoundException(key);
+                    }
+                    return value;
+                }
+            }
+
+            public bool TryGetValue(string key, [MaybeNullWhen(false)] out Type value)
+            {
+                var lookup = table.Lookup(TypeHashingAlgorithms.ComputeNameHashCode(key));
+                NativeParser entryParser;
+                while (!(entryParser = lookup.GetNext()).IsNull)
+                {
+                    string foundName = entryParser.GetString();
```

Comment:
Is this materializing a string for every check in a given chain? 🤔

Reply:
Found another method that doesn't do that (just wasn't exposed here) and switched to that.